yjit/src/codegen.rs

```diff
@@ -5762,23 +5762,75 @@ fn jit_rb_str_empty_p(
     return true;
 }
 
-// Codegen for rb_str_concat() -- *not* String#concat
-// Frequently strings are concatenated using "out_str << next_str".
-// This is common in Erb and similar templating languages.
-fn jit_rb_str_concat(
+// Codegen for rb_str_concat() with an integer argument -- *not* String#concat
+// Using strings as a byte buffer often includes appending byte values to the end of the string.
+fn jit_rb_str_concat_codepoint(
     jit: &mut JITState,
     asm: &mut Assembler,
     _ci: *const rb_callinfo,
     _cme: *const rb_callable_method_entry_t,
     _block: Option<BlockHandler>,
     _argc: i32,
     _known_recv_class: Option<VALUE>,
+) -> bool {
+    asm_comment!(asm, "String#<< with codepoint argument");
+
+    // Ensure the codepoint argument is a Fixnum.
+    let arg = asm.stack_opnd(0);
+    let comptime_arg = jit.peek_at_stack(&asm.ctx, 0);
+    if comptime_arg.fixnum_p() {
+        jit_guard_known_klass(
+            jit,
+            asm,
+            comptime_arg.class_of(),
+            arg,
+            arg.into(),
+            comptime_arg,
+            SEND_MAX_DEPTH,
+            Counter::guard_send_not_fixnums,
+        );
+    } else {
+        return false;
+    }
```

Comment:
Here you already checked that `comptime_arg` is a fixnum in the caller.

Maybe we can also fold the guard that the value is fixnum in the caller? Because presumably the fallback `rb_str_concat(str, codepoint)` can handle any kind of input type? @XrXr would this be valid?

Reply:
The original version with the assembler required the type to be a `Fixnum`, but that's no longer the case. I can move this guard easily enough. I just want to confirm that if we see another type at the call site we don't want to try to cut over to `jit_rb_str_concat`?

Reply:
If we're going to allow non-codepoint arguments into `rb_yjit_str_concat_codepoint`, should I rename that function? Or is naming it after its primary intention okay?

Reply:
> I just want to confirm that if we see another type at the call site we don't want to try to cut over to jit_rb_str_concat?

Seems like if it's already behind a `FIXNUM_P` check in the caller (checking in a different way), so you could turn this into an assert.

Renaming the C function sounds good. Maybe `rb_yjit_str_concat_likely_byte`? Also a short comment on the C side would be nice.

Reply:
I was looking at this with @paracycle, too. Since we're talking about turning `rb_yjit_str_concat_codepoint` into a generic fallback, should we collapse the `jit_rb_str_concat` and `jit_rb_str_concat_codepoint` codegen methods into a single handler that calls to this C function?

I don't know the history of `jit_rb_str_concat` to know whether it's doing more in assembly for speed or because of necessity. That's what I used as the template for my original implementation of `jit_rb_str_concat_codepoint` and we decided to move things out to C to keep the inline code block smaller and reduce branching. If we applied those same principals here, I suppose we could simplify the codegen a fair bit. But, if that's out of scope, that's fine with me.

Reply:
Collapsing them into a single C call sounds good to me. Most of today's `jit_rb_str_concat` seems like something that could be done in the C function.

Reply:
I've consolidate the two methods and pushed that up. It's ready for review if anyone is interested.